code/core/src/phone/java/com/adobe/marketing/mobile/services/ui/notification/PushTemplateImageUtil.kt

```diff
@@ -83,10 +53,105 @@ internal object PushTemplateImageUtil {
      *
      * @param cacheService the AEPSDK [CacheService] to use for caching or retrieving
      * downloaded image assets
-     * @param uri [String] containing an image asset url
+     * @param uriList [String] containing an image asset url
      * @return [Bitmap] containing the image referenced by the `String` uri
      */
-    internal fun downloadImage(cacheService: CacheService, uri: String?): Bitmap? {
+    internal fun downloadImage(
+        cacheService: CacheService,
+        uriList: List<String?>
+    ): Int {
+        val assetCacheLocation = getAssetCacheLocation()
+        if (assetCacheLocation.isNullOrEmpty() || uriList.isEmpty()) {
+            return 0
+        }
+
+        var downloadedImageCount = 0
+        val latch = CountDownLatch(uriList.size)
+        for (uri in uriList) {
+            if (uri == null || !UrlUtils.isValidUrl(uri)) {
+                latch.countDown()
+                continue
+            }
+
+            val cacheResult = cacheService[assetCacheLocation, uri]
+            if (cacheResult != null) {
+                Log.trace(
+                    PushTemplateConstants.LOG_TAG,
+                    SELF_TAG,
+                    "Found cached image for $uriList"
+                )
+                downloadedImageCount++
+                latch.countDown()
+                continue
+            }
+
+            download(uri) { image ->
+                image?.let {
+                    Log.trace(
+                        PushTemplateConstants.LOG_TAG,
+                        SELF_TAG,
+                        "Successfully download image from $uriList"
+                    )
+                    downloadedImageCount++
+
+                    // scale down the bitmap to 300dp x 200dp as we don't want to use a full
+                    // size image due to memory constraints
+                    val pushImage = scaleBitmap(image)
+                    // write bitmap to cache
+                    try {
+                        bitmapToInputStream(pushImage).use { bitmapInputStream ->
+                            cacheBitmapInputStream(
+                                cacheService,
+                                bitmapInputStream,
+                                uri
+                            )
+                        }
+                    } catch (exception: IOException) {
+                        Log.trace(
+                            PushTemplateConstants.LOG_TAG,
+                            SELF_TAG,
+                            "Exception occurred creating an input stream from a bitmap: ${exception.localizedMessage}."
+                        )
+                    }
+                }
+                latch.countDown()
+            }
+        }
+        latch.await()
```

Comment:
should the await call have a time limit to prevent a hang if network issues occur?

Reply:
Good point, I've added timeout of `DOWNLOAD_TIMEOUT_SECS`